federation/pkg/controller/service/servicecontroller.go

```diff
@@ -0,0 +1,827 @@
+		return false
+	}
+
+	if x.Port != y.Port {
+		return false
+	}
+
+	if x.NodePort != y.NodePort {
+		return false
+	}
+
+	// We don't check TargetPort; that is not relevant for load balancing
+	// TODO: Should we blank it out?  Or just check it anyway?
+
+	return true
+}
+
+func intSlicesEqual(x, y []int) bool { // TODO: quinton - this belongs in the utils library, not here.  Move it.
+	if len(x) != len(y) {
+		return false
+	}
+	if !sort.IntsAreSorted(x) {
+		sort.Ints(x)
+	}
+	if !sort.IntsAreSorted(y) {
+		sort.Ints(y)
+	}
+	for i := range x {
+		if x[i] != y[i] {
+			return false
+		}
+	}
+	return true
+}
+
+func stringSlicesEqual(x, y []string) bool { // TODO: quinton - this belongs in the utils library, not here.  Move it.
+	if len(x) != len(y) {
+		return false
+	}
+	if !sort.StringsAreSorted(x) {
+		sort.Strings(x)
+	}
+	if !sort.StringsAreSorted(y) {
+		sort.Strings(y)
+	}
+	for i := range x {
+		if x[i] != y[i] {
+			return false
+		}
+	}
+	return true
+}
+
+func hostsFromServiceList(list *api.ServiceList) []string { // TODO: quinton: This probably needs to return IP's or DNS names, not "hostnames".  Rename and recode.
+	result := []string{}
+	for ix := range list.Items {
+		if list.Items[ix].Spec.Unschedulable { // TODO: quinton: In the case of cluster services this relates to the health of the services, not the nodes.  Fix this.
+			continue
+		}
+		result = append(result, list.Items[ix].Name)
+	}
+	return result
+}
+
+//TODO: quinton: Add a good comment/doc
+func getClusterServiceConditionPredicate() cache.ClusterServiceConditionPredicate {
+	return func(clusterService kube_api.Service) bool {
+		if clusterService.Status.Health != OK { //TODO: quinton: Fix the health check logic here.
+			return false
+		}
+		// If we have no info, don't accept
+		if len(clusterService.Status.Conditions) == 0 { // TODO: quinton: Check whether this applies
+			return false
+		}
+		for _, cond := range clusterService.Status.Conditions {
+			// We consider the cluster service for inclusion in DNS records only when its ServiceReady condition status
+			// is ConditionTrue
+			// TODO: quinton: Check up on exactly how service status works, and fix.
+			if cond.Type == api.ServiceReady && cond.Status != api.ConditionTrue {
+				glog.V(4).Infof("Ignoring cluster service %v with %v condition status %v", clusterService.Name, cond.Type, cond.Status)
+				return false
+			}
+		}
+		return true
+	}
+}
+
+// clusterServiceSyncLoop handles updating the load balancers pointed to by all federated service DNS Records
+// whenever the set of cluster services in the federation changes.
+// TODO: quinton: Assumed that cluster service load balancers have host (i.e. DNS) names (and IP's?).  Verify and fix as necessary.
+//                Also the logic here bundles all hos/DNS names of all services in the federation together - need to handle service one separately.  Rewrite.
+func (s *ServiceController) clusterServiceSyncLoop(period time.Duration) {
```

Comment:
@quinton-hoole, about the last statement "if there is newly added cluster, add all services to it." in #23848 . My previous idea is use a informercontroller to watch the change of clusters, if there is new cluster being added, then add all services to it in dns server, and if there is a cluster being removed, then removed all entries related to this cluster. 

But this depends on how the dns provider api is designed. I found the current cloud provider api need all hosts to be passed in, if dns provider follow the same, then I do not expect a rewrite on this function... as we always need to pass all current available clusters into the api.
```
EnsureLoadBalancer(service *api.Service, hosts []string, annotations map[string]string) (*api.LoadBalancerStatus, error)
UpdateLoadBalancer(service *api.Service, hosts []string) error
```